homeassistant/components/omnilogic/sensor.py

```diff
@@ -0,0 +1,485 @@
+                            PERCENT_UNITS,
+                            bow["CSAD"],
+                        )
+                    )
+
+    async_add_entities(sensors, update_before_add=True)
+
+
+class OmnilogicSensor(OmniLogicEntity):
+    """Defines an Omnilogic sensor entity."""
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        kind: str,
+        name: str,
+        backyard: dict,
+        bow: dict,
+        device_class: str,
+        icon: str,
+        unit: str,
+        sensordata: dict,
+    ):
+        """Initialize Entities."""
+        super().__init__(
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            backyard=backyard,
+            bow=bow,
+            icon=icon,
+            sensordata=sensordata,
+        )
+        self._state = None
+        self._unit_type = backyard["Unit-of-Measurement"]
+        self._device_class = device_class
+        self._unit = None
+
+    @property
+    def should_poll(self) -> bool:
+        """Return the polling requirement of the entity."""
+        return False
+
+    @property
+    def device_class(self):
+        """Return the device class of the entity."""
+        if self._device_class != "none":
+            return self._device_class
+
+    @property
+    def unit_of_measurement(self):
+        """Return the right unit of measure."""
+        return self._unit
+
+    @property
+    def force_update(self):
+        """Force update."""
+        return True
+
+    @property
+    def state(self):
+        """Return the state."""
+
+        _LOGGER.debug("Updating state of sensor: %s", self._name)
+        if self._kind == "water_temperature":
+            sensordata = None
+
+            for backyard in self.coordinator.data:
+                for bow in backyard["BOWS"]:
+                    if bow["systemId"] == self.bow["systemId"]:
+                        sensordata = bow["waterTemp"]
+                        break
+
+            temp_return = int(sensordata)
+            temp_state = int(sensordata)
+            unit_of_measurement = TEMP_FAHRENHEIT
+            if self._backyard["Unit-of-Measurement"] == "Metric":
+                temp_return = round((temp_return - 32) * 5 / 9, 1)
+                unit_of_measurement = TEMP_CELSIUS
+
+            if int(sensordata) == -1:
+                temp_return = None
+                temp_state = None
+
+            self._attrs["hayward_temperature"] = temp_return
+            self._attrs["hayward_unit_of_measure"] = unit_of_measurement
+            self._state = temp_state
+            self._unit = TEMP_FAHRENHEIT
+            self._attrs["systemId"] = self.bow.get("systemId")
+            self._name = (
+                self._backyard["BackyardName"]
+                + " "
+                + self.bow.get("Name")
+                + " Water Temperature"
+            )
+
+        elif self._kind == "filter_pump_speed":
```

Comment:
There seems to be several distinct type of sensors. Maybe better to use subclasses instead of branching logic?

Reply:
This will be done later today.

Reply:
Addressed.